File: completed/0440.java
import java.util.*;

class Solution {
    private int count(long prefix1, long prefix2, int max) {
        int steps = 0;
        while (prefix1 <= max) {
            steps += Math.min(max + 1, prefix2) - prefix1;
            prefix1 *= 10;
            prefix2 *= 10;
        }
        return steps;
    }

    public int findKthNumber(int n, int k) {
        int current = 1;
        k--;
        while (k > 0) {
            int steps = count(current, current + 1, n);
            if (steps <= k) {
                k -= steps;
                current++;
            } else {
                k--;
                current *= 10;
            }
        }
        return current;
    }
}

File: completed/0440.cpp
class Solution {
private:
    int inBetween(long current, long next, long n) {
        int count = 0;
        while (current <= n) {
            count += min(n + 1, next) - current;
            current *= 10;
            next *= 10;
        }
        return count;
    }

public:
    int findKthNumber(int n, int k) {
        int current = 1;
        k--;
        while (k > 0) {
            int count = inBetween(current, current + 1, n);
            if (count <= k) {
                k -= count;
                current++;
            } else {
                k--;
                current *= 10;
            }
        }
        return current;
    }
};
